### DISCUSSION TITLE
New Tree Sitter Parser for Emacs Lisp

### DISCUSSION THREAD

**Discussion by Wilfred** [tree-sitter/tree-sitter#1344] [Grammars]
I've written a pretty complete parser for emacs lisp (elisp) using tree-sitter!

https://github.com/Wilfred/tree-sitter-elisp

I've been amazed at how easy it was to write and test. It works for 99.87% of the elisp programs I've tested it on!

The only major issue I've had is parsing files that contain null bytes. Is this supported? `/\0/` complains about back references and `'\0'` has no effect as far as I can see.

**Comment by sogaiu** [comment]
I think null bytes didn't used to work.  Not sure what the current status is.  I'm also interested in being able to support that in some grammars.

**Comment by ahlinc** [comment]
~The `\0` null bytes [don't allowed](https://github.com/tree-sitter/tree-sitter/issues/1167) in tree-sitter grammars, they have a special purpose for tree-sitter itself and serve as EOF~ _(**edit:** see discussion bellow)_. Also such bytes [shouldn't be used](https://github.com/AbstractMachinesLab/tree-sitter-erlang/pull/10) in regexes.

I suppose there were no concerns about this limitation before, but if it's a noticeable problem I think this limitation may be eliminated in the future.

Cause I don't clearly understand why `\0` bytes affect text files and why they are present in such files, @Wilfred could you explain why `\0` bytes exists in the `elisp` source files? Does it a common situation and what purpose of the `\0` bytes in elisp source files?

**Reply by sogaiu** [reply]
FWIW, I have found them in Clojure source code ([in the wild](https://github.com/santifa/clj-dbase/blob/a269ca62d529cf82cec7bffce2e38b71458c6087/src/clj_dbase/core.clj#L121)) and they are also [legal in Janet source code](https://github.com/janet-lang/janet/blob/e552757edcaf8805deec82cdf740504e8f563785/src/core/parse.c#L38) (treated as whitespace).

**Reply by ahlinc** [reply]
@sogaiu thanks for the note, in my understanding those NUL bytes are included into the mentioned grammars as resistance measure against malformed text files or _(I don't know for sure)_ such languages use different from C language convention for strings representation like `\0` terminated strings. In my understanding it's generally unusual that NUL bytes exist in text files.

**Reply by sogaiu** [reply]
@ahlinc Thanks for your thoughts.

I think it's likely as you say that using `\0` in source code is not very common (but this is a guess).

At least in the Clojure source file that I linked to, there is a literal `\0` -- it's right after the backslash (if you clone the repository and take a look at the file with a hex editor, it can be observed directly).

IIUC, the code in question [there](https://github.com/santifa/clj-dbase/blob/a269ca62d529cf82cec7bffce2e38b71458c6087/src/clj_dbase/core.clj#L121) `(complement #{\\0})` is saying "everything except `\0`" -- so AFAICT, the `\0` is deliberately included.  (Please see the end of this comment for an explanation about `\\0`.)

It may be as you say that in Janet `\0` is treated as whitespace for better handling.  I asked before about why but did not receive a response.

For both languages, based on testing against largish samples of source files, I do not think it's common to use `\0` in source files -- I just know that they are handled.

For reference, here are a few of the issues I've noticed where `\0` was mentioned before:

* #280 
* https://github.com/tree-sitter/tree-sitter-ocaml/pull/24
* #475 

At the end of #475 I saw:

> Soon, (once Atom has been upgraded to use the newest version of the library), I'll change the default behavior of tree-sitter generate to rely on the new API.

I don't know if that ever happened.

<details>
<summary>About `\\0`</summary>
I put the `\0` there so it's visible.  In my text above it looks like there are 3 characters `\\0`, but actually the intent is 2.  The first backslash indicates the beginning of something for Clojure to understand as a character, and the following byte (`\0`) is which character ( cf . https://github.com/sogaiu/tree-sitter-clojure/blob/39bf0977d223879436c1425fe6bfeb3bcfd86f92/grammar.js#L173-L183 )
</details>

**Reply by ahelwer** [reply]
Heh, I've been using `lexer->lookahead == 0` to detect EOF in my external scanner, time to switch to the API call I guess. I recall trying it out once and it led to weird behavior but I must have been doing something else wrong. Another thing is that I think calling `lexer->advance` when you're at EOF is a no-op, right? Is it also a no-op when it's at a null character?

**Reply by maxbrunsfeld** [reply]
Yeah, I believe that null bytes should be supported now, and that Tree-sitter no longer assumes that they indicate EOF.

**Reply by Wilfred** [reply]
Yeah, I've seen null bytes in the wild too. The file I hit was https://github.com/realgud/realgud/blob/master/realgud/common/bp-image-data.el, which has a TIFF file stored in a string literal. Interestingly GitHub doesn't render the file at all.

@maxbrunsfeld OK, after a little experimentation, it looks like I can do things like `choice("foo", "\0")` but I haven't found a way of matching null bytes in regexps. Is it currently possible?

**Reply by ahlinc** [reply]
@sogaiu thank you for detailed description and interesting links, I didn't see those before.

@maxbrunsfeld thank you for enlightening your position about the NUL bytes case, there is an issue #1167 about current situation with NUL bytes.

@Wilfred ~It's not possible for now~ to use `"\0"` in the tree-sitter grammars #1167 and in regexps also: https://github.com/AbstractMachinesLab/tree-sitter-erlang/pull/10, for now just the best thing is just to avoid to specify it in the grammar and replace such bytes to the `\x20` spaces in a buffer before passing it to the tree-sitter parser.

**Reply by ahelwer** [reply]
Could also convert the regex to an external token, right? Does it also work inside calls to `token`?

**Reply by ahelwer** [reply]
After some testing (the hardest part of which was actually typing a null character) I can confirm that if you rewrite your regex to be some `choice`/`seq`/`repeat` functions inside a `token` function, it will match null characters:
```js
rules: {
  source_file: $ => token(repeat1(seq('a', '\0')))
}
```
When run against a file consisting of `aNULLaNULLaNULLaNULL` I get the debug output:
>new_parse
process version:0, version_count:1, state:1, row:1, col:0
lex_internal state:0, row:1, column:0
  consume character:'a'
  consume character:0
  consume character:'a'
  consume character:0
  consume character:'a'
  consume character:0
  consume character:'a'
  consume character:0
lexed_lookahead sym:source_file_token1, size:8
shift state:2
process version:0, version_count:1, state:2, row:1, col:8
lex_internal state:0, row:1, column:8
lexed_lookahead sym:end, size:0
reduce sym:source_file, child_count:1
accept
done
(source_file [0, 0] - [0, 8])

Also tested that `token` doesn't allow any extras to occur between its members, so it really does seem to function as a sort of DSL for writing regexes (don't need to use `token.immediate` or anything).

**Reply by ahlinc** [reply]
Hmm, interesting, @ahelwer thanks for the check, it looks that the `'\0'` specified by the string literal works for builtin tree-sitter lexer but at the same time in the external scanners the `lexer->lookahead == 0` construction means a different thing and currently equals to the `lexer->is_eof()` or not?

**Reply by ahelwer** [reply]
`lexer->lookahead == 0` is confirmed different from `lexer->eof(lexer)`; this will correctly parse the same file: https://github.com/ahelwer/tree-sitter-test/blob/2dd3c048897043a06268ef92a400b29b818aac0b/src/scanner.cc#L40
> new_parse
process version:0, version_count:1, state:1, row:1, col:0
lex_external state:1, row:1, column:0
  consume character:'a'
  consume character:0
  consume character:'a'
  consume character:0
  consume character:'a'
  consume character:0
  consume character:'a'
  consume character:0
lexed_lookahead sym:token, size:8
shift state:2
process version:0, version_count:1, state:2, row:1, col:8
lex_internal state:0, row:1, column:8
lexed_lookahead sym:end, size:0
reduce sym:source_file, child_count:1
accept
done
(source_file [0, 0] - [0, 8]
  (token [0, 0] - [0, 8]))

Of course there is a danger here; `lexer->eof(lexer)` implies `lexer->lookahead == 0`, although not vice-versa. Thus if you are trying to consume an arbitrary number of null characters and don't check whether you are at EOF, you will enter an infinite loop because calling `lexer->advance` while at EOF is a no-op.

**Reply by ahlinc** [reply]
@ahelwer Thank you a lot for checking all those corner cases!!!
Now the picture is much more clear and above findings definitely need to be documented.

**Reply by ahelwer** [reply]
I was thinking I'd like to massively expand the documentation on external scanners.
